doc/proposals/helm-operator.md

```diff
@@ -0,0 +1,127 @@
+* Create a new watches file at `./<project-name>/watches.yaml`. The chart and GVK will be defaulted based on input to the `new` command.
+* Create a `./<project-name>/deploy` with the Kubernetes resource files necessary to run the operator.
+* Create a `./build/Dockerfile` that uses the watches file and the helm chart. It will use the Helm operator as its base image.
+
+The resulting structure will be:
+
+```
+<project-name>
+|   watches.yaml
+|
+|-- build
+|   |   Dockerfile
+|
+|-- charts
+|   |-- <kind>
+|       |   Chart.yaml
+|       |   ...
+|
+|-- deploy
+|   |   operator.yaml
+|   |   role_binding.yaml
+|   |   role.yaml
+|   |   service_account.yaml
+|   |
+|   |-- crds
+|       |   <gvk>_crd.yaml
+|       |   <gvk>_cr.yaml
+```
+
+#### Add
+
+Add functionality will be updated to allow Helm operator developers to add new CRDs/CRs and to update the watches.yaml file for additional Helm charts. The command helps when a user wants to watch more than one CRD for their operator.
+
+```
+operator-sdk add api --api-version=<group>/<version> --kind=<kind>
+```
+
+Flags:
+* **Required:** --kind - the kind for the CRD.
+* **Required:** --api-version - the group/version for the CRD.
+
+**NOTE:** `operator-sdk add controller` will not be supported, since it doesn't make sense for a Helm operator.
+
+#### Up
+
+Up functionality will be updated to allow Helm operator developers to run their operator locally, using the `operator-sdk` binary's built-in helm operator implementation.
+
+```
+operator-sdk up local
+```
+
+This should use the known structure and the helm operator code to run the operator from this location. The existing code will need to be updated with a new operator type check for `helm` (in addition to existing `go` and `ansible` types). The command works by running the operator-sdk binary, which includes the Helm operator code, as the operator process.
+
+#### Build
+
+Build functionality will be updated to support building a docker image from the Helm operator directory structure.
+
+```
+operator-sdk build <image-name>
+```
+
+### Observations and open questions
+
+* There will be a large amount of overlap in the `operator-sdk` commands for the Ansible and Helm operators. We should take care to extract the resusable features of the Ansible operator commands into a shared library, usable by both Helm and Ansible commands.
+
+* This proposal assumes that a Helm Operator base image will be available for building Helm operator projects. What generates the Helm operator base image and what is the registry, image name, versioning, etc.?
+
+* There is a moderate amount of complexity already related to how operator types are handled between the `go` and `ansible` types. With the addition of a third type, there may need to be a larger design proposal for operator types. For example, do we need to define an `Operator` interface that each of the operator types can implement for flag verification, scaffolding, project detection, etc.?
```

Comment:
I do think there is complexity, I would prefer to keep the complexity as close to the surface. Something like a factory pattern which each up command(helm, go, ansible) have their own implementation and their own flags possibly? 

I also would like to at least have an idea on what to do here in this proposal IMO just so we don’t live in a super complex world while we figure this out? Thoughts?

Reply:
@shawn-hurley Yeah that's pretty much what I was thinking. I'm not sure what the cleanest way to do that would be though.

Maybe we could do the operator type detection / `--type` flag parsing before passing things off to cobra and then have each operator type implement the cobra subcommand functions directly or pass the operator implementation to the subcommand functions?

Reply:
> Maybe we could do the operator type detection / --type flag parsing before passing things off to cobra and then have each operator type implement the cobra subcommand functions directly or pass the operator implementation to the subcommand functions

This seems like we'll have implicit subcommands:
- `operator-sdk build --type=go` ==> `operator-sdk go build`
- `operator-sdk build --type=ansible` ==> `operator-sdk ansible build`

This goes back to the question of what sub commands should be allowed in what types of projects, and how explicit should we be about specifying the type. #670 

Some commands are common across different types(`build`, `up local`) whereas other sub commands only make sense for certain project types e.g Go(`add api`, `generate k8s`) Ansible/Helm(`add crd`). It gets a little confusing if we want to allow those exclusive commands across all types to allow migrating to a hybrid Go/Ansible Go/Helm operator.

One option is we stay with the existing convention of specifying the `--type` flag(or inferring from the project) for each subcommand so we can keep the commands consistent for all 3 types, and handle the logic for each type within the same sub command:
- `operator-sdk up local`(infer the type) or `operator-sdk up local --type=<type>`

This keeps all subcommands consistent, except we still have to specify the --type flag or infer that from the project. The downside is the ambiguity in what commands are allowed where.

Or we take a top down approach as @joelanford suggested (https://github.com/operator-framework/operator-sdk/issues/670#issuecomment-434835370) and put the operator type as the first root sub command:
- `operator-sdk go build`
- `operator-sdk ansible build`
- `operator-sdk helm build`

The advantage of the type root subcommand is that there's no ambiguity on what commands run in what types of projects with differing arguments and flags.
The drawback is that it's a little verbose(redundant?) for commands that are (mostly)common in their behavior
e.g `operator-sdk build`.
Also if I end up with a hybrid operator do I run `operator-sdk go build` or `operator-sdk ansible build`?

Reply:
Great points. Maybe its a combination. I think we can categorize the CLI into three sets of commands.

1) Global commands where operator type does not need to be inferred:
    * `completion`, `help`, and `new`

2) Commands that need to support possible hybrid operators and where the operator type(s) will need to be inferred:
    * `up`, `build`, `test`

3) Commands that apply to specific types:
    * Go: `add api`, `add controller`, `generate`
    * Ansible: `add crd`
    * Helm: ?

For 2., as @estroz mentioned (https://github.com/operator-framework/operator-sdk/issues/670#issuecomment-435107379), we can detect all of the types an operator has and be opinionated about which combinations we'll support, with helpful errors/warnings about non-supported combinations.

For 3., another option may be to keep `add` and `generate` as top-level commands, but then have type subcommands of those. So:
* `operator-sdk add go api`
* `operator-sdk generate go k8s`
* `operator-sdk add ansible crd` (or maybe even better, `operator-sdk add ansible role`)
* `operator-sdk add helm crd` (or `add helm chart`)

Reply:
> For 3., another option may be to keep add and generate as top-level commands, but then have type subcommands of those. So:

I was thinking of something along those lines as well. Would make things clear when you do `operator-sdk add go --help` as it would display all the subcommands.